text/0000-unsized-expr.md

```diff
@@ -0,0 +1,216 @@
+freely if they want a `Trait` to be returned in a `Rc<Trait>`, `Arc<Trait>`, or
+`Box<Trait>`. They can also decide if they want a slice to be appended to a
+vector or if they want a completely new vector to be allocated.
+
+## New concepts
+
+It's also interesting to see that this syntax allows us to express new concepts
+in safe Rust that could previously only be expressed with unsafe code.
+
+Consider the following concept:
+
+** The class of types that can be turned into a slice **
+
+With the proposed syntax, this can be expressed as follows:
+
+```rust
+trait IntoSlice<T> {
+    fn into_slice(self) -> [T];
+}
+
+impl<'a, T: Copy> IntoSlice<T> for &'a [T] {
+    fn into_slice(self) -> [T] { self[..] }
+}
+
+impl<T> IntoSlice<T> for Box<[T]> {
+    fn into_slice(self) -> [T] { *self }
+}
+```
+
+It is not possible to have one trait that encapsulates both of these
+implementations safely without the `[T]` syntax. Let's see why:
+
+First of all, since we want the trait to be implemented for all `Box<[T]>`, not
+just those `T` that implement `Copy`, the trait has to take `self` by value.
+Since the return value has to work for `&'a [T]`, the return value has to be
+`&'a [T]`. Therefore the trait has to look like this:
+
+```rust
+trait IntoSlice<'a, T> {
+    fn into_slice(self) -> &'a [T];
+}
+```
+
+But this cannot be implemented for `Box<[T]>` because we either leak the heap
+allocation or return a reference to an already deallocated slice.
+
+(For similar reasons it's not possible to implement the trait on
+`&mut Box<[T]>`)
+
+# Detailed design
+
+Allow the return types of functions to be unsized.
+
+Allow the type of an expression to be unsized if it is *founded in* a `box` or
+`in` expression and the allocator accepts the unsized type.
+
+Not all allocators accept all unsized types, e.g., `Box` can hold both slices
+and traits but `Vec` can only hold slices.
+
+An expression is *founded in* a `box` or `in` expression if it is either used
+directly in a `box` or `in` expression, or if it is used as the value of an
+expression which is founded in a `box` or `in` expression, or if it is used as
+the return value of a function.
+
+Some examples of valid expressions and functions:
+
+```rust
+box *"";
+box { *"" };
+box { { *"" } };
+
+fn f(n: usize) -> str {
+    if n == 0 {
+        *""
+    } else {
+        f()
+    }
+}
+```
+
+The following implementation is not necessarily good but is documented here to
+show the feasibility of the concept:
+
+## Implementation
+
+If a function returns an unsized type, it is passed a hidden pointer to the
+allocator which is used to store the return value. Once the return value has
+been determined, the function asks the allocator to allocate an appropriate
+amount of memory and stores the return value in there. The allocator also
+receives the necessary metadata: The length of the slice for slices and the
+vtable pointer for traits.
+
+# Drawbacks
+
+None right now.
+
+# Alternatives
+
+None right now.
```

Comment:
You mean the pointer to the allocator? That would mean that the function has to manually deal with the allocator at every return point. While the interface suggested in the `box/in` RFC isn't explicitly marked unsafe, it's actually extremely unsafe. If the user requests too little memory or doesn't initialize the memory properly, the behavior is most likely undefined. I don't think this is a real alternative.